### PR TITLE
Option to add contract scope for k8s services as an annotation

### PR DESCRIPTION
**Pull request by apoorva11029**
Default value: context

For example, add in the service definition:
"metadata": {
  "annotations": {
    "opflex.cisco.com/ext_service_scope" : "tenant"
  }
}

Valid scope values: context, tenant, global. An empty entry defaults to context(VRF only)

If a new service is deployed with invalid scope, it wont register with APIC
If an existing service is updated with invalid scope, the service wont be updated in APIC

Added unit and functional tests to check the contract scope values and service annotations.
Also formatted the committed files with go fmt

### CODE REVIEW

File: pkg/metadata/annotations.go [hunk 1]
```diff
@@ -46,6 +46,9 @@ type NetIps struct {
 // Service endpoint annotation
 const ServiceEpAnnotation = "opflex.cisco.com/service-endpoint"
 
+// Service scope annotation
+const ServiceScopeAnnotation = "opflex.cisco.com/ext_service_scope"
```
Comment (snaiksat) [inline pkg/metadata/annotations.go]: I  am wondering if we should call this ServiceContractScopeAnnotation, and also update the annotation string to:
opflex.cisco.com/ext_service_contract_scope

to make things less ambiguous. Also the comment should be a little more descriptive, indicating what is the relevance of this annotation, how it should be used, what are the legal values, and what is the default. I would recommend reflecting this information in the commit message as well.

File: pkg/controller/services_test.go [hunk 2]
```diff
@@ -185,33 +183,242 @@ func TestServiceIp(t *testing.T) {
...
+	_, err = apicContract(name, tenant, graphName, "wrong-Name")
+	assert.NotNil(t, err)
+
+	contract, err = apicContract(name, tenant, graphName, "")
+	assert.Empty(t, err)
+	assert.Equal(t, contract.GetAttrStr("scope"), "context")
+
+	contract, err = apicContract(name, tenant, graphName, "GloBal")
+	assert.Empty(t, err)
+	assert.Equal(t, contract.GetAttrStr("scope"), "global")
```
Comment (snaiksat) [inline pkg/controller/services_test.go]: Is there an assert needed for the third legal value "tenant"?

File: pkg/metadata/annotations.go (hunk 1 above)
Comment (apoorva11029) [reply pkg/metadata/annotations.go]: Makes sense, I will add this to code and commit.

File: pkg/controller/services_test.go (hunk 2 above)
Comment (apoorva11029) [reply pkg/controller/services_test.go]: That's to make sure a blank scope value("") defaults to "context"

File: pkg/controller/services.go [hunk 3]
```diff
@@ -325,15 +325,40 @@ func apicExtNetCons(conName string, tenantName string,
...
+	return true
+}
 func apicContract(conName string, tenantName string,
-	graphName string) apicapi.ApicObject {
+	graphName string, scopeName string) (apicapi.ApicObject, error) {
+	scopeName_norm := strings.ToLower(scopeName)
+	checkScope := validScope(scopeName_norm)
+	if !checkScope {
+		return nil, errors.New("InvalidServiceContractScope")
+	}
```
Comment (mchalla) [inline pkg/controller/services.go]: could you add the scope name when  returning this error.

File: pkg/controller/cf_update_handlers.go [hunk 4]
```diff
@@ -341,8 +343,11 @@ func (env *CfEnvironment) createAppServiceGraph(appId string, extIps []string) a
 			apicExtNet(name, cont.config.AciVrfTenant,
 				cont.config.AciL3Out, extIps))
 
-		serviceObjs = append(serviceObjs,
-			apicContract(name, cont.config.AciVrfTenant, graphName))
+		contract, err := apicContract(name, cont.config.AciVrfTenant, graphName, conScope)
+		if err != nil {
+			panic(err.Error())
```
Comment (jojimt) [inline pkg/controller/cf_update_handlers.go]: Let's replace the panic with an error return. Please also open an issue to track the need for validators for this annotation.

File: pkg/controller/services.go [hunk 5]
```diff
@@ -397,11 +422,16 @@ func (cont *AciController) updateServiceDeviceInstance(key string,
 		nodes = append(nodes, node)
 	}
 	sort.Strings(nodes)
-
 	name := cont.aciNameForKey("svc", key)
+	var conScope string
+	scope_val, ok := service.ObjectMeta.Annotations[metadata.ServiceScopeAnnotation]
+	if ok {
+		conScope = scope_val
```
Comment (mchalla) [inline pkg/controller/services.go]: I assume ok could also mean its empty "" ?

File: pkg/controller/services.go [hunk 6]
```diff
@@ -325,15 +325,40 @@ func apicExtNetCons(conName string, tenantName string,
...
+		if v == str {
+			return true
+		}
+	}
+	return false
+}
+
+func validScope(scope string) bool {
+	validValues := []string{"", "context", "tenant", "global"}
+	if !stringInSlice(scope, validValues) {
```
Comment (jojimt) [inline pkg/controller/services.go]: Please change to return stringInSlice(..) directly.

File: pkg/controller/services.go [hunk 7]
```diff
@@ -325,15 +325,40 @@ func apicExtNetCons(conName string, tenantName string,
...
+		return false
+	}
+	return true
+}
 func apicContract(conName string, tenantName string,
-	graphName string) apicapi.ApicObject {
+	graphName string, scopeName string) (apicapi.ApicObject, error) {
+	scopeName_norm := strings.ToLower(scopeName)
```
Comment (jojimt) [inline pkg/controller/services.go]: nit: please stick to just camel case. scopeName_norm can be renamed as normScopeName.

File: pkg/controller/services.go (hunk 3 above)
Comment (jojimt) [reply pkg/controller/services.go]: Can we also avoid declaring checkScope? You can directly check the function return value.

File: pkg/controller/services.go (hunk 5 above)
Comment (apoorva11029) [reply pkg/controller/services.go]: If it's empty, we default to "context", but empty string is still acceptable

File: pkg/controller/services.go [hunk 8]
```diff
@@ -397,11 +422,16 @@ func (cont *AciController) updateServiceDeviceInstance(key string,
...
 	}
 	sort.Strings(nodes)
-
 	name := cont.aciNameForKey("svc", key)
+	var conScope string
+	scope_val, ok := service.ObjectMeta.Annotations[metadata.ServiceScopeAnnotation]
+	if ok {
+		conScope = scope_val
+	} else {
+		conScope = "context"
```
Comment (jojimt) [inline pkg/controller/services.go]: Please define a constant named defaultContext and use it here.

File: pkg/controller/services.go [hunk 9]
```diff
@@ -916,7 +951,10 @@ func (cont *AciController) handleServiceUpdate(service *v1.Service) bool {
 		cont.indexMutex.Lock()
 		if cont.serviceSyncEnabled {
 			cont.indexMutex.Unlock()
-			cont.updateServiceDeviceInstance(servicekey, service)
+			err = cont.updateServiceDeviceInstance(servicekey, service)
+			if err != (interface{})(nil) {
```
Comment (jojimt) [inline pkg/controller/services.go]: No need for a typecast here. err != nil is sufficient.

File: pkg/controller/services.go [hunk 10]
```diff
@@ -15,25 +15,27 @@
...
+	"sort"
+	"strconv"
+	"strings"
 
 	"github.com/noironetworks/aci-containers/pkg/apicapi"
 	"github.com/noironetworks/aci-containers/pkg/metadata"
 )
 
+var defaultServiceContractScope = "context"
+
```
Comment (jojimt) [inline pkg/controller/services.go]: Please define this as const instead of var.

**Pull request merged**